### PR TITLE
Add store credits

### PR DESCRIPTION
**Pull request by JDutil**
This is an updated branch in favor of https://github.com/spree/spree/pull/5625
- [x] Frontend UX
- [x] Frontend UX Feature Specs
- [x] Backend UX for creating credit
- [x] Backend UX for creating credit reasons
- [ ] Final code review & houndci cleanup.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://add-store-credits.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://add-store-credits.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: core/app/models/spree/store_credit.rb [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    end
+
+    def credit_allocation_memo
+      "This is a credit from store credit ID #{self.id}"
+    end
+
+    def store_event
+      return unless amount_changed? || amount_used_changed? || amount_authorized_changed? || action == ELIGIBLE_ACTION
+
+      event = nil
```
Comment (forvalho) [inline core/app/models/spree/store_credit.rb]: Would'nt it make more sense like this?

```
event = if action
          store_credit_events.build(action: action)
        else
          store_credit_events.where(action: ALLOCATION_ACTION).first_or_initialize
        end
```

**Comment by MattDunbar** [comment]
I feel like this would be better suited for its own gem.

Across many Spree builds we've done as an agency I'm yet to come across one that actually needs this, and building payment methods into the core seems a bit much while trying to keep it lightweight.

Although I'd lean towards this being an entirely separate gem in a separate repo, if its agreed that this needs to be part of the base it'd be great to isolate this and the existing check payment method into their own gem as part of this repo still.

**Comment by athal7** [comment]
@MattDunbar it already is an extension on spree-contrib. It was decided that the store-credit part of that extension would be useful for core, and the virtual gift card part of that extension is being left off for now until there appears to be more desire.

**Comment by peterberkenbosch** [comment]
@MattDunbar @athal7 while trying to get this merged into core, I tend to feel the same Matt. It's actually a lot of code, and also has some limitations currently (only works with a creditcard payment method). Will discuss this further today. @JDutil wdyt.

**Comment by athal7** [comment]
You can remove the credit card restriction, that is Bonobos logic creeping in. Let me know if a chat would be beneficial or if you want me to take a look at certain areas that are concerning to you.

**Comment by peterberkenbosch** [comment]
Yes let's chat!

**Comment by paultyng** [comment]
This still has the issue of depending on presence of a confirm step.  You can see our workarounds in [my comment in the previous issue](https://github.com/spree/spree/pull/5625#issuecomment-69361879).  Besides removing dependency on the optional step though, I think our changes simplified the addition of the payments by removing the need to do the reconciliation with existing payments.

I'd be happy to chat in the gitter room about our workarounds we've implemented as well to get the extension on our prod site if you are interested, I wish I had some time to contribute on this one as I'd really like to get to stop maintaining this part of our codebase, but I probably won't free up for a couple weeks yet, if then.

**Comment by nimashariatian** [comment]
just curious if this is going to be merged soon?

**Comment by peterberkenbosch** [comment]
Yes it will. Some minor issues to finish.

On Mon, Mar 2, 2015 at 11:10 PM, nimashariatian <notifications@github.com>
wrote:

> just curious if this is going to be merged soon?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spree/spree/pull/5917#issuecomment-76837422>.
>



-- 

Peter Berkenbosch
Spree Commerce, Inc.

File: core/app/models/spree/store_credit.rb [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+      if (amount_used + amount_authorized) > amount
+        errors.add(:amount_authorized, Spree.t('admin.store_credits.errors.amount_authorized_exceeds_total_credit'))
+      end
+    end
+
+    def validate_no_amount_used
+      if amount_used > 0
+        errors.add(:amount_used, 'is greater than zero. Can not delete store credit')
+      end
+    end
```
Comment (iangreenleaf) [inline core/app/models/spree/store_credit.rb]: This method should return false, or else the destroy will go ahead. You could do it like this:

```ruby
def validate_no_amount_used
  if amount_used > 0
    errors.add(...')
  end
  errors.blank?
end
```

Comment (JDutil) [inline core/app/models/spree/store_credit.rb]: Feel free to submit a PR against the store credit branch to fix this.

**Comment by johnwook** [comment]
Thanks for the great works, all! I think merging is under pending because of some conflicts. Anyone that can handle merge conflicts, or is there anything that I can do to make this happen? :) :+1:

**Comment by JDutil** [comment]
@johnwook you could submit a PR against add-store-credits branch that merges master branch into it, and/or improve/review/test the PR descriptions task list.

File: backend/app/controllers/spree/admin/store_credits_controller.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+      def load_categories
+        @credit_categories = Spree::StoreCreditCategory.all.order(:name)
+      end
+
+      def load_store_credit
+        @store_credit = Spree::StoreCredit.find_by_id(params[:id]) || Spree::StoreCredit.new
+      end
+
+      def ensure_unused_store_credit
```
Comment (jasonfb) [inline backend/app/controllers/spree/admin/store_credits_controller.rb]: two things: 1) this is domain logic and shouldn't be in a controller. (But that is forgivable)

2) We actually have a specific need from our Customer Service team that store credits be adjustable after they are created. In our implementation, we enforce that you cannot lower the store credit amount to below the used + authorized amounts.

File: core/app/models/spree/payment.rb [hunk 4]
```diff
@@ -30,6 +30,7 @@ class Payment < Spree::Base
 
     # invalidate previously entered payments
     after_create :invalidate_old_payments
+    after_create :create_eligible_credit_event
```
Comment (jasonfb) [inline core/app/models/spree/payment.rb]: is there a way we can do this without callbacks? Overuse of callbacks is large problem in Spree, and should be curtailed

File: backend/app/controllers/spree/admin/store_credits_controller.rb (hunk 3 above)
Comment (jasonfb) [inline backend/app/controllers/spree/admin/store_credits_controller.rb]: also... raising errors is no good user experience for my Admins. Better to show them pretty errors, regardless of how we go with the functionality.

**Comment by jasonfb** [comment]
@johnwook @JDutil 

At Mack Weldon we have a compelling need to get this into Spree 3.1 (we are currently using our own SC implementation and I would like to merge it with this implementation when we upgrade), so I can probably put some dev resources on this within the next 3 week timeframe. 

I left some minor feedback ... since so much work has gone into this already, I would not consider my feedback to be show-stopper for this moving forward as-is (since 1, the feedback is minor, and 2, those things could be addressed in subsequent releases)

Perhaps all it needs to be squashed & rebased?

**Comment by jasonfb** [comment]
The only other 'feature' we need is to be able to expire store credits--- this is *very* important to our stakeholders, as it affects the amount of debt on our books. That, too, is a feature which I think can be added later, and one in which I can assign some developer resources to building out.

**Comment by JDutil** [comment]
@jasonfb I don't think a squash rebase is necessary this is pretty large branch, and I'm fine with it just being merged to maintain history.

I believe there was the ability to expire store credits in this PR already, but they are non-expiring by default. It's been awhile since I've reviewed this though.

**Comment by jasonfb** [comment]
Re: Git Process.. I will defer to you (in my app I always squash & rebase because I hate merge commits). 

Re: Expiration....  Unless I've missed something, assuming this implementation is based on bonobos/spree-store-credit-as-payment-method (which it looks like, with some small differences), confusingly, there is an "expiring" and "non-expiring" *type* --- and there is some calculus to apply the expiring ones before the non-expiring ones, but the *actual expiration* is not implemented. in other words, some of the code is written anticipation of expiration, but actual expiration is not implemented (this is also true for bonobos/spree-store-credit-as-payment-method, in which the author stated simply that the expiration was a planned future extension)

You can see I'm right just but looking at the store credits table itself --- it has no expires_at field.

**Comment by JDutil** [comment]
Usually I prefer the squash/rebase, but with large PRs like this it's best to keep the history.

This implementation was based on the bonobos extension, but has had a different frontend for it implemented as they just used the API. 

They may have finished working on expiration in solidus or maybe not I'm not really sure.

**Comment by johnwook** [comment]
@JDutil @jasonfb Thanks for replies! I'll try my best to make this branch merged. Also, I'm looking forward to seeing @jasonfb 's pre-implemented store credit functions including expiration.

**Comment by jasonfb** [comment]
@johnwook --- in our app (which uses similar code lifted from bonobos/store_credit_as_payment_method) we did the expiration in a really wonky way that I'm not particularly happy with. It basically runs a cron job and on the day of expiration, reset the ```amount``` of the store credit to be whatever you haven't yet used (either 0 or the amount used+amount authorized)

it's not very elegant, so I wouldn't even recommend it for the core.

**Comment by justindrees** [comment]
@johnwook would love to have this functionality for a site I'm working on. Any update on when/if this will be merged? Many thanks!

**Comment by johnwook** [comment]
Hi @justindrees, sorry for late reply.
Recently, I moved from srpee to [solidus](solidus.io), and solidus has built in store-credit functionality.
Check some articles why I moved to solidus.

- [The Future of Spree Open Source Software](https://spreecommerce.com/blog/future-of-spree-oss)
- [The Future of Spree and the Future of Solidus](https://solidus.io/blog/2015/10/28/future-of-spree.html)

It took a couple of days for me to complete the migration(spree 3.0 branch -> solidus)

**Comment by Mafi88** [comment]
@damianlegawiec @priyank-gupta Could you please check if everything is ok here?

Don't correct code according to: ```Use && instead of and```. There should be ```and``` as it is flow operator, ```&&``` will break changing path.

**Pull request merged**